tpm docs (#3071)
hashicorp/web-unified-docs: content/vault/v2.x/content/api-docs/secret/identity/tpm.mdx

---
layout: api
page_title: 'Identity Secret Backend: TPM - HTTP API'
description: This is the API documentation for managing TPM records in the identity store.
# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2026-06-25T00:00:00.000Z
last_modified: 2026-06-25T00:00:00.000Z
# END AUTO GENERATED METADATA
---

<EnterpriseAlert />

# Identity secrets engine - TPM (API)

A TPM (Trusted Platform Module) record in Vault's identity store represents a
physical or virtual TPM device identified by its Endorsement Key (EK) public
key. TPM records allow Vault to track and control access for specific TPM

> devices.

Reviewer: ```suggestion
Vault represents a physical or virtual Trusted Platform Module (TPM) device as a
record in the identity store identified by its Endorsement Key (EK) public key.
TPM records allow Vault to track and control access for specific TPM devices.
```
Style correction: avoid possessives

Key characteristics of TPM records:

- The **ID** is automatically derived as the SHA-256 hash of the decoded bytes
  from the `tpm_ek_public_key` field and cannot be changed after creation.
- The `tpm_ek_public_key` field is **immutable** — it can only be set at

>   creation time.

Reviewer: ```suggestion
- The `tpm_ek_public_key` field is **immutable** and is only set at creation time.
```
Style correction: avoid em dashes

- TPM IDs are scoped to a namespace, so the same EK public key may be
  registered in multiple namespaces without conflict.
- A **disabled** TPM cannot be used for authentication or certificate
  generation.

## Create a TPM

This endpoint creates a new TPM record. The `tpm_ek_public_key` field is
required and must be a PEM-encoded public key with the `PUBLIC KEY` header.

> 

Reviewer: ```suggestion
Create a new TPM record from a PEM-encoded public key with the `PUBLIC KEY`
header.
```
It's okay if we want to give folks a heads up that they need some info handy, but the parameter definition should already make it clear which fields are/are not required

If a `name` is not provided, Vault generates a default name in the format

> `tpm_<random-hex>`.

Reviewer: ```suggestion
```
Folded into the name field definition

| Method | Path             |
| :----- | :--------------- |
| `POST` | `/identity/tpm`  |

### Parameters

- `name` `(string: "tpm-<random-hex>")` – A human-friendly name for the TPM
  record. Must be unique within the namespace. Can be updated after creation

>   using the update endpoints.

Reviewer: ```suggestion
- `name` `(string: "tpm-<random-hex>")` – A human-friendly name for the TPM
  record. The TPM name must be unique within the namespace. If you do not provide
  a name, Vault generates a default name in the format `tpm_<random-hex>`. You
  can update the name later using the update endpoint.
```
Style correction: use complete sentences, write in active voice

- `tpm_ek_public_key` `(string: <required>)` – PEM-encoded TPM Endorsement Key
  (EK) public key. Must be a valid PEM block with type `PUBLIC KEY`. This value

>   is **immutable** — it cannot be changed after the TPM record is created.

Reviewer: ```suggestion
- `tpm_ek_public_key` `(string: <required>)` – PEM-encoded TPM Endorsement Key
  (EK) public key. The key must be a valid PEM block with type `PUBLIC KEY`. The
  key value is **immutable** and cannot be changed after the TPM record is created.
```

- `metadata` `(key-value-map: {})` – Metadata to be associated with the TPM
  record. In the CLI, this parameter can be repeated multiple times and all

>   values are merged together.

Reviewer: ```suggestion
  record. In CLI commands, you can repeat the `metadata` parameter multiple
  times and Vault will merge them together to create a single update string.
```

- `disabled` `(bool: false)` – Whether the TPM is disabled. A disabled TPM

>   cannot generate certificates or be used for authentication.

Reviewer: ```suggestion
- `disabled` `(bool: false)` – Whether the TPM is disabled. You cannot generate
  certificates or authenticate with a disabled TPM.
```
Style correction: write in active voice

### Sample payload

```json
{
  "name": "my-tpm",
  "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n",
  "metadata": {
    "machine_name": "server-01",
    "location": "datacenter-east"
  },
  "disabled": false
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request POST \
    --data @payload.json \
    ${VAULT_ADDR}/v1/identity/tpm
```

### Sample response

```json
{
  "data": {
    "disabled": false,
    "id": "sha256-a3f1c2d4e5b6...",
    "metadata": {
      "location": "datacenter-east",
      "machine_name": "server-01"
    },
    "name": "my-tpm",
    "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n"
  }
}
```

## Read TPM by ID

> This endpoint queries a TPM record by its identifier.

Reviewer: ```suggestion
Query a TPM record by its identifier.
```

| Method | Path                    |
| :----- | :---------------------- |
| `GET`  | `/identity/tpm/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM record.

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    ${VAULT_ADDR}/v1/identity/tpm/id/sha256-a3f1c2d4e5b6...
```

### Sample response

```json
{
  "data": {
    "disabled": false,
    "id": "sha256-a3f1c2d4e5b6...",
    "metadata": {
      "location": "datacenter-east",
      "machine_name": "server-01"
    },
    "name": "my-tpm",
    "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n"
  }
}
```

## Update TPM by ID

> This endpoint updates an existing TPM record by its identifier.

Reviewer: ```suggestion
Update an existing TPM record by its identifier.
```

The `tpm_ek_public_key` field cannot be changed. Supplying a value that differs
from the existing key returns an error.

| Method | Path                    |
| :----- | :---------------------- |
| `POST` | `/identity/tpm/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM record.

- `name` `(string: <optional>)` – Updated name for the TPM record. Must be

>   unique within the namespace.

Reviewer: ```suggestion
- `name` `(string: <optional>)` – Updated name for the TPM record. The new name
  must be unique within the namespace.
```

- `metadata` `(key-value-map: {})` – Updated metadata. Replaces the existing
  metadata entirely.

- `disabled` `(bool: false)` – Whether the TPM is disabled.

### Sample payload

```json
{
  "name": "renamed-tpm",
  "metadata": {
    "machine_name": "server-01-updated"
  },
  "disabled": true
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request POST \
    --data @payload.json \
    ${VAULT_ADDR}/v1/identity/tpm/id/sha256-a3f1c2d4e5b6...
```

### Sample response

```json
{
  "data": {
    "disabled": true,
    "id": "sha256-a3f1c2d4e5b6...",
    "metadata": {
      "machine_name": "server-01-updated"
    },
    "name": "renamed-tpm",
    "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n"
  }
}
```

## Delete TPM by ID

This endpoint deletes a TPM record by its identifier. Deleting a record that

> does not exist is idempotent and returns no error.

Reviewer: ```suggestion
Delete a TPM record by its identifier. eleting a record that does not exist is
idempotent and returns successfully as a no-op.
```

| Method   | Path                    |
| :------- | :---------------------- |
| `DELETE` | `/identity/tpm/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM record.

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request DELETE \
    ${VAULT_ADDR}/v1/identity/tpm/id/sha256-a3f1c2d4e5b6...
```

## List TPMs by ID

This endpoint returns a list of all TPM record identifiers in the current

> namespace.

Reviewer: ```suggestion
Fetch a list of all TPM record identifiers in the current namespace.
```

| Method | Path                          |
| :----- | :---------------------------- |
| `LIST` | `/identity/tpm/id`            |
| `GET`  | `/identity/tpm/id?list=true`  |

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request LIST \
    ${VAULT_ADDR}/v1/identity/tpm/id
```

### Sample response

```json
{
  "data": {
    "keys": [
      "sha256-a3f1c2d4e5b6...",
      "7d8e9f0a1b2c..."
    ]
  }
}
```

## Create/Update TPM by name

This endpoint creates or updates a TPM record by name. If no TPM with the
given name exists, a new record is created (and `tpm_ek_public_key` is

> required). If a record with that name already exists, it is updated.

Reviewer: ```suggestion
Create or update a TPM record by name. Vault creates a new record if no TPM with
the given name exists and the `tpm_ek_public_key` is set. Otherwise, Vault
updates the record with the given name, if it exists.
```
Style correction: write in active voice

| Method | Path                        |
| :----- | :-------------------------- |
| `POST` | `/identity/tpm/name/:name`  |

### Parameters

- `name` `(string: <required>)` – Name of the TPM record.

- `tpm_ek_public_key` `(string: <required on create>)` – PEM-encoded TPM
  Endorsement Key public key. Required when creating a new record. Cannot be

>   changed on an existing record.

Reviewer: ```suggestion
  Endorsement Key public key. Vault requires the key when creating a new record.
  You cannot change the key on an existing record.
```

- `metadata` `(key-value-map: {})` – Metadata to be associated with the TPM
  record. Replaces the existing metadata entirely on update.

- `disabled` `(bool: false)` – Whether the TPM is disabled.

### Sample payload

```json
{
  "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n",
  "metadata": {
    "machine_name": "server-02"
  }
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request POST \
    --data @payload.json \
    ${VAULT_ADDR}/v1/identity/tpm/name/my-tpm
```

### Sample response

```json
{
  "data": {
    "disabled": false,
    "id": "sha256-a3f1c2d4e5b6...",
    "metadata": {
      "machine_name": "server-02"
    },
    "name": "my-tpm",
    "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n"
  }
}
```

## Read TPM by name

> This endpoint queries a TPM record by its name.

Reviewer: ```suggestion
Query a TPM record by its name.
```

| Method | Path                        |
| :----- | :-------------------------- |
| `GET`  | `/identity/tpm/name/:name`  |

### Parameters

- `name` `(string: <required>)` – Name of the TPM record.

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    ${VAULT_ADDR}/v1/identity/tpm/name/my-tpm
```

### Sample response

```json
{
  "data": {
    "disabled": false,
    "id": "sha256-a3f1c2d4e5b6...",
    "metadata": {
      "machine_name": "server-02"
    },
    "name": "my-tpm",
    "tpm_ek_public_key": "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkq...\n-----END PUBLIC KEY-----\n"
  }
}
```

## Delete TPM by name

This endpoint deletes a TPM record by its name. Deleting a record that does

> not exist is idempotent and returns no error.

Reviewer: ```suggestion
Delete a TPM record by its name. Deleting a record that does not exist is
idempotent and returns successfully as a no-op.
```

| Method   | Path                        |
| :------- | :-------------------------- |
| `DELETE` | `/identity/tpm/name/:name`  |

### Parameters

- `name` `(string: <required>)` – Name of the TPM record.

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request DELETE \
    ${VAULT_ADDR}/v1/identity/tpm/name/my-tpm
```

## List TPMs by name

> This endpoint returns a list of all TPM record names in the current namespace.

Reviewer: ```suggestion
Fetch a list of all TPM record names in the current namespace.
```
